mypy/constant_fold.py

```diff
@@ -0,0 +1,116 @@
+"""Constant folding of expressions.
+
+For example, 3 + 5 can be constant folded into 8.
+"""
+
+from __future__ import annotations
+
+from typing import Union
+from typing_extensions import Final
+
+from mypy.nodes import Expression, FloatExpr, IntExpr, NameExpr, OpExpr, StrExpr, UnaryExpr, Var
+
+# All possible result types of constant folding
+ConstantValue = Union[int, bool, float, str]
+CONST_TYPES: Final = (int, bool, float, str)
+
+
+def constant_fold_expr(expr: Expression, cur_mod_id: str) -> ConstantValue | None:
+    """Return the constant value of an expression for supported operations.
+
+    Among other things, support int arithmetic and string
+    concatenation. For example, the expression 3 + 5 has the constant
+    value 8.
+
+    Also bind simple references to final constants defined in the
+    current module (cur_mod_id). Binding to references is best effort
+    -- we don't bind references to other modules. Mypyc trusts these
+    to be correct in compiled modules, so that it can replace a
+    constant expression (or a reference to one) with the statically
+    computed value. We don't want to infer constant values based on
+    stubs, in particular, as these might not match the implementation
+    (due to version skew, for example).
+
+    Return None if unsuccessful.
+    """
+    if isinstance(expr, IntExpr):
+        return expr.value
+    if isinstance(expr, StrExpr):
+        return expr.value
+    if isinstance(expr, FloatExpr):
+        return expr.value
+    elif isinstance(expr, NameExpr):
+        if expr.name == "True":
+            return True
+        elif expr.name == "False":
+            return False
```

Comment:
Interesting, this is very similar logic to the code @JelleZijlstra added in `evalexpr.py` in https://github.com/python/mypy/commit/a9c62c5f82f34a923b8117a5394983aefce37b63. I wonder if that logic could be reused here?

Reply:
Yeah, there is definitely some overlap. Right now the benefits of sharing code don't seem big enough to me to make it worthwhile to increase coupling, but if we add support for more things, it may make sense to refactor and share parts of the implementations. Note that mypy and mypyc already share some of the constant folding code.

Reply:
Makes sense!

Reply:
Should this be a visitor?

Reply:
I'm not using a visitor, since only relatively few AST node types are handled here, and others have shared default behavior. I generally use a visitor when I need to implement logic for many node types. However, I think that in this case both using and not using a visitor would have been reasonable.